### PR TITLE
AdminMergeDevelopment

### PR DESCRIPTION
**Pull request by nole23**
Ostala jedna metoda koja nije uradjena za admina iz razloga sto mora se sacekati da se uradi entity Tenant.

### CODE REVIEW

File: src/main/java/com/buildingmanagementapp/models/Supervisor.java [hunk 1]
```diff
@@ -21,7 +21,7 @@
 @Entity
 @Table(name = "supervisors")
 @Inheritance(strategy = InheritanceType.JOINED)
-public abstract class Supervisor implements ISupervisor {
+public class Supervisor implements ISupervisor {
```
Comment (foo4foo) [inline src/main/java/com/buildingmanagementapp/models/Supervisor.java]: Why is abstract keyword removed ?

File: src/main/java/com/buildingmanagementapp/models/User.java [hunk 2]
```diff
@@ -61,23 +60,17 @@
 	private boolean activated = false;
 
 	@ManyToMany
-	  @JoinTable(
-	      name="user_role",
-	      joinColumns=@JoinColumn(name="user_id", referencedColumnName="id"),
-	      inverseJoinColumns=@JoinColumn(name="role_id", referencedColumnName="id"))
+	@JoinTable(name = "user_role", joinColumns = @JoinColumn(name = "user_id", referencedColumnName = "id"), inverseJoinColumns = @JoinColumn(name = "role_id", referencedColumnName = "id"))
```
Comment (foo4foo) [inline src/main/java/com/buildingmanagementapp/models/User.java]: This line is longed than 80 characters.

File: src/main/java/com/buildingmanagementapp/services/AdminService.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+	private boolean admin(Principal principal) {
+		Set<Role> role = userRepo.findByUsername(principal.getName())
+				.getRoles();
+
+		for (Role o : role) {
+
+			// System.out.println(o.getName());
+			if (!o.getName().equals("ADMIN"))
+				return false;
```
Comment (foo4foo) [inline src/main/java/com/buildingmanagementapp/services/AdminService.java]: So you go through user roles to find an admin. If there are more than one role, the first one is not an admin, and the second one is, this method will return false even though user is an admin.

File: src/main/java/com/buildingmanagementapp/services/AdminService.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+		supervisorRepo.save(supervisor);
+
+		return HttpStatus.OK;
+
+	}
+
+	public HttpStatus saveFirm(FirmDTO firmDTO, Principal principal, Long id,
+			Long addres, String type) {
+
+		if (admin(principal) == false)
```
Comment (foo4foo) [inline src/main/java/com/buildingmanagementapp/services/AdminService.java]: Method 'admin' already returns boolean. It is redundant to do it here again.

File: src/main/java/com/buildingmanagementapp/services/AdminService.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+			return HttpStatus.NOT_FOUND;
+
+		if (user.getRoles().isEmpty()) {
+			user.getRoles().add(roleRepo.findByName("ADMIN"));
+		} else {
+			for (Role o : user.getRoles()) {
+
+				user.getRoles().remove(roleRepo.findByName(o.getName()));
+			}
+			user.getRoles().add(roleRepo.findByName("ADMIN"));
```
Comment (foo4foo) [inline src/main/java/com/buildingmanagementapp/services/AdminService.java]: Deleting all roles and adding admin role again if it exists? Why ?

File: src/main/java/com/buildingmanagementapp/services/AdminService.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+		supervisorRepo.delete(supervisor.getId());
+
+		Set<Firm> firms = new HashSet<Firm>();
+		for (User u : firm.getUsers()) {
+			u.setFirms(firms);
+		}
+
+		firmRepo.delete(firm);
+
+		return HttpStatus.ACCEPTED;
```
Comment (foo4foo) [inline src/main/java/com/buildingmanagementapp/services/AdminService.java]: HTTP 202 ACCEPTED - The request has been accepted for processing, but the processing has not been completed. The request might or might not be eventually acted upon, and may be disallowed when processing occurs.
This method did its job, so it should return OK or NO_CONTENT.

File: src/main/java/com/buildingmanagementapp/services/AdminService.java [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+			return HttpStatus.NOT_FOUND;
+		
+		for (Role o : user.getRoles()) {
+
+			user.getRoles().remove(roleRepo.findByName(o.getName()));
+		}
+		user.getRoles().add(roleRepo.findByName("EMPLOYEE"));
+		
+		return HttpStatus.ACCEPTED;
+	}
```
Comment (foo4foo) [inline src/main/java/com/buildingmanagementapp/services/AdminService.java]: Same here.

File: src/main/java/com/buildingmanagementapp/controllers/AdminController.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+ * Create any method for AdminController.
+ * - Add building. Method = POST and Mapping "/admin/add/building/{id_admin}"
+ * - Add firm. Method = POST and Mapping "/admin/add/firm/{id_user}/{address}/{type}"
+ * - Add admin of system. Method = PUT and Mapping "/add/system/admin/{id_user}"
+ * - Delete firm. Method = DELETE and Mapping "/delet/firm/{id_firm}"
+ * - Delete admin. Method = PUT and Mapping "/delet/system/admin/{id_user}"
+ * - Get all user. Method = GET and Mapping "/all/user"
+ * - Get all building. Method = GET and Mapping "/all/building"
+ * - Get all firm. Method = GET and Mapping "/all/firm"
+ *
```
Comment (foo4foo) [inline src/main/java/com/buildingmanagementapp/controllers/AdminController.java]: Use plural :dancing_women:

File: src/main/java/com/buildingmanagementapp/services/AdminService.java (hunk 3 above)
Comment (nole23) [reply src/main/java/com/buildingmanagementapp/services/AdminService.java]: This is my logic error. I'll fix this.

File: src/main/java/com/buildingmanagementapp/services/AdminService.java (hunk 4 above)
Comment (nole23) [reply src/main/java/com/buildingmanagementapp/services/AdminService.java]: This is my sintax error.

File: src/main/java/com/buildingmanagementapp/services/AdminService.java (hunk 5 above)
Comment (nole23) [reply src/main/java/com/buildingmanagementapp/services/AdminService.java]: I was using a unique role for users.

File: src/main/java/com/buildingmanagementapp/services/AdminService.java (hunk 6 above)
Comment (nole23) [reply src/main/java/com/buildingmanagementapp/services/AdminService.java]: I didn't know. I'll fix this.

File: src/main/java/com/buildingmanagementapp/services/AdminService.java [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+
+		return HttpStatus.OK;
+	}
+
+	public HttpStatus deleteFirm(Long id, Principal principal) {
+		// TODO Auto-generated method stub
+
+		if (admin(principal))
+			return HttpStatus.UNAUTHORIZED;
+
```
Comment (foo4foo) [inline src/main/java/com/buildingmanagementapp/services/AdminService.java]: Same here.

File: src/main/java/com/buildingmanagementapp/controllers/AdminController.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+	}
+
+	/**
+	 * Delete firm
+	 * 
+	 * @return HttpStatus.UNAUTHORIZED or NOT_FOUND or ACCEPTER
+	 * 
+	 */
+	@RequestMapping(value = "/delet/firm/{id_firm}", method = RequestMethod.DELETE)
+	public ResponseEntity<String> deleteFirm(@PathVariable Long id_firm,
```
Comment (foo4foo) [inline src/main/java/com/buildingmanagementapp/controllers/AdminController.java]: Maybe delete ?

**Pull request merged**